docs/decisions/0012-semantic-function-multiple-model-support.md

```diff
@@ -0,0 +1,260 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: {proposed}
+date: 2023-10-02
+deciders: mabolan, markwallace
+consulted: semenshi, dmytrostruk, rbarreto
+informed: 
+---
+# Multiple Model Support for Semantic Functions
+
+## Context and Problem Statement
+
+Developers need to be able to use multiple models e.g., using chat completion together with embeddings.
+
+<!-- This is an optional element. Feel free to remove. -->
+## Use Cases
+
+* Select Model Request Settings by Default Service Id
+* Select Model Request Settings by Model Id
+* Select Model By Semantic Function Preference
+* Select Model By Developer Defined Strategy
+* Select Model Request Settings By Developer Defined Strategy
+
+## Decision Outcome
+
+Chosen option: "{title of option 1}", because
+{justification. e.g., only option, which meets k.o. criterion decision driver | which resolves force {force} | … | comes out best (see below)}.
+
+## Descriptions of the Use Cases
+
+### Select Model Request Settings by Default Service Id
+
+_As a developer using the Semantic Kernel I can configure multiple request settings for a semantic function and associate each one with a service id so that the correct request settings are used when different services are used to execute my semantic function._
+
+The semantic function template configuration allows multiple model request settings to be configured. In this case the developer configures different settings based on the service id that is used to execute the semantic function.
+In the example below the semantic function is executed with "OpenAIChat" using `max_tokens=240` because "OpenAIChat" is set as the default AI service.
+
+```csharp
+// Configure a Kernel with multiple LLM's
+IKernel kernel = Kernel.Builder
+    .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+    .WithAzureTextCompletionService(deploymentName: aoai.DeploymentName,
+        endpoint: aoai.Endpoint, serviceId: "AzureText", apiKey: aoai.ApiKey)
+    .WithAzureChatCompletionService(deploymentName: aoai.ChatDeploymentName,
+        endpoint: aoai.Endpoint, serviceId: "AzureChat", apiKey: aoai.ApiKey)
+    .WithOpenAITextCompletionService(modelId: oai.ModelId,
+        serviceId: "OpenAIText", apiKey: oai.ApiKey)
+    .WithOpenAIChatCompletionService(modelId: oai.ChatModelId,
+        serviceId: "OpenAIChat", apiKey: oai.ApiKey, setAsDefault: true)
+    .Build();
+
+// Configure semantic function with multiple LLM request settings
+string configPayload = @"{
+  ""schema"": 1,
+  ""description"": ""Hello AI, what can you do for me?"",
+  ""models"": [
+    { ""service_id"": ""AzureText"", ""max_tokens"": 60 },
```

Comment:
IMO, configurations should be agnostic of the `Service Id` but not from the `Service Type`.

Reason, plugins can be imported from other sources (with configurations that should be common to everyone). Using `types` instead of `ids` defined by the dev.

```
models: [ 
    { 
        service_type: "IChatCompletion", "max_tokens": 400
        // This one will accept any IChatCompletion available
    }, 
    { 
        service_type: "AzureChatCompletion", "max_tokens": 200
        // This one will override the above if the IChatCompletion setup is a `AzureChatCompletion` type.
    }
]

```

Reply:
I'm not sure if naming `service_type: "IChatCompletion"` or `service_type: "AzureChatCompletion"` will align with Python or Java implementations, since it looks like `IChatCompletion` and `AzureChatCompletion` are C# interfaces/types. The name should be compatible for all SK versions, either it's `Service Id` or `Service Type`.

While I understand the purpose of `Service Id`, it would be nice to learn more about `Service Type` approach and its benefits.

Reply:
The goal of using `service_id` is to provide an unambiguous link between the request settings and the service i.e., these are the settings for service `foobar`. There are drawbacks with this approach e.g., if you want to create a prompt that works in multiple environments but for that use case there are other solutions in the ADR.

Reply:
> I'm not sure if naming service_type: "IChatCompletion" or service_type: "AzureChatCompletion" will align with Python or Java implementations,

You are right, we might consider having a global ServiceType and ServiceName across platforms.

Completion example:
Category: IChatCompletion -> "chat_completion"
Name: AzureChatCompletion -> "azure_chat_completion"

Memory example:
Category: IMemoryStore -> "memory_store"
Name: AzureCognitiveSearchMemoryStore-> "azure_cognitive_search_memory_store"

Suggestion:
```csharp
public interface IAIService 
{
    string Category { get; }
    string Name { get; }
}
```

Json:
```
models: [ 
    { 
        service_category: "chat_completion", "max_tokens": 400
        // This one will accept any IChatCompletion available
    }, 
    { 
        service_name: "azure_chat_completion", "max_tokens": 200
        // This one will override the above if the IChatCompletion setup is a `AzureChatCompletion` type.
    }
]
```

Reply:
What you have called "Service Name" is the same as "Service Id". This is existing functionality so there is no need to change the name.

The proposal is to use "Model Id" rather than service category because it is more specific e.g., the request settings for GPT 3.5 and GPT 4 may be different event though both at chat completion services.

This ADR is specific to Request Settings for AI Models and Memory and other services are out of scope.

Reply:
Switching based on model id and service type is out of scope for this ADR